dig/node.go

```diff
@@ -0,0 +1,131 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package dig
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/pkg/errors"
+)
+
+type graphNode interface {
+	// Return value of the object
+	value(g *graph) (reflect.Value, error)
+
+	// Other things that need to be present before this object can be created
+	dependencies() []interface{}
+}
+
+type node struct {
+	objType     reflect.Type
+	cached      bool
+	cachedValue reflect.Value
+}
+
+type objNode struct {
+	node
+	fmt.Stringer
+
+	obj interface{}
+}
+
+// Return the earlier provided instance
+func (n objNode) value(g *graph) (reflect.Value, error) {
+	return reflect.ValueOf(n.obj), nil
+}
+
+func (n objNode) dependencies() []interface{} {
+	return nil
+}
+
+func (n objNode) String() string {
+	return fmt.Sprintf(
+		"(object) obj: %v, deps: nil, cached: %v, cachedValue: %v",
+		n.objType,
+		n.cached,
+		n.cachedValue,
+	)
+}
+
+type funcNode struct {
+	node
+	fmt.Stringer
+
+	constructor interface{}
+	deps        []interface{}
+}
+
+// Call the function and return the result
+func (n *funcNode) value(g *graph) (reflect.Value, error) {
+	if n.cached {
+		return n.cachedValue, nil
+	}
+
+	cv := reflect.ValueOf(n.constructor)
```

Comment:
yes, but any time you write reflect related function, you almost always need type and value, one or multiple times. So i got into a habit of doing that at the top, i don't see much harm in that